Detectors/TPC/spacecharge/src/PoissonSolver.cxx

```diff
@@ -0,0 +1,1497 @@
+    const DataT tempFourth = 1.0 / (2.0 + 2.0 * tempRatio);
+    const DataT inverseTempFourth = 1.0 / tempFourth;
+    calcCoefficients2D(1, tnRRow - 1, h, coefficient1, coefficient2);
+    Vector matricesCurrentV = tvArrayV[count - 1];
+    Vector matricesCurrentCharge = tvCharge[count - 1];
+    Vector residue = tvResidue[count - 1];
+
+    // 1) Pre-Smoothing: Gauss-Seidel Relaxation or Jacobi
+    for (int jPre = 1; jPre <= nPre; ++jPre) {
+      relax2D(matricesCurrentV, matricesCurrentCharge, tnRRow, tnZColumn, h2, tempFourth, tempRatio, coefficient1, coefficient2);
+    }
+
+    // 2) Residue calculation
+    residue2D(residue, matricesCurrentV, matricesCurrentCharge, tnRRow, tnZColumn, ih2, inverseTempFourth, tempRatio, coefficient1, coefficient2);
+
+    iOne = 2 * iOne;
+    jOne = 2 * jOne;
+    tnRRow = iOne == 1 ? Nr : Nr / iOne + 1;
+    tnZColumn = jOne == 1 ? Nz : Nz / jOne + 1;
+
+    matricesCurrentCharge = tvCharge[count];
+    matricesCurrentV = tvArrayV[count];
+
+    //3) Restriction
+    restrict2D(matricesCurrentCharge, residue, tnRRow, tnZColumn, 0);
+  }
+
+  // Do V cycle from: gridTo-1 to gridTo gamma times
+  for (int iGamma = 0; iGamma < gamma; ++iGamma) {
+    vCycle2D(gridTo - 1, gridTo, nPre, nPost, gridSizeR, ratio, tvArrayV, tvCharge, tvResidue);
+  }
+
+  // Go to finest grid
+  for (int count = gridTo - 2; count >= gridFrom; --count) {
+    iOne = iOne * 0.5;
+    jOne = jOne * 0.5;
+
+    const DataT h = gridSizeR * iOne;
+    const DataT h2 = h * h;
+    const DataT tempRatio = ratio * iOne * iOne / (jOne * jOne);
+    const DataT tempFourth = 1.0 / (2.0 + 2.0 * tempRatio);
+
+    tnRRow = iOne == 1 ? Nr : Nr / iOne + 1;
+    tnZColumn = jOne == 1 ? Nz : Nz / jOne + 1;
+    const Vector matricesCurrentCharge = tvCharge[count - 1];
+    Vector matricesCurrentV = tvArrayV[count - 1];
+    const Vector matricesCurrentVC = tvArrayV[count];
+
+    // 6) Interpolation/Prolongation
+    addInterp2D(matricesCurrentV, matricesCurrentVC, tnRRow, tnZColumn, 0);
+
+    calcCoefficients2D(1, tnRRow - 1, h, coefficient1, coefficient2);
+
+    // 7) Post-Smoothing: Gauss-Seidel Relaxation
+    for (Int_t jPost = 1; jPost <= nPost; ++jPost) {
+      relax2D(matricesCurrentV, matricesCurrentCharge, tnRRow, tnZColumn, h2, tempFourth, tempRatio, coefficient1, coefficient2);
+    } // end post smoothing
+  }
+}
+
+template <typename DataT, size_t Nz, size_t Nr, size_t Nphi>
+void PoissonSolver<DataT, Nz, Nr, Nphi>::vCycle2D(const int gridFrom, const int gridTo, const int nPre, const int nPost, const DataT gridSizeR, const DataT ratio, std::vector<Vector>& tvArrayV,
+                                                  std::vector<Vector>& tvCharge, std::vector<Vector>& tvResidue)
+{
+  int iOne = 1 << (gridFrom - 1);
+  int jOne = 1 << (gridFrom - 1);
+
+  int tnRRow = iOne == 1 ? Nr : Nr / iOne + 1;
+  int tnZColumn = jOne == 1 ? Nz : Nz / jOne + 1;
+
+  std::vector<DataT> coefficient1(Nr);
```

Comment:
I am not quite getting the rational why the GRID size is a template argument everywhere. But since it is, there is no need to use dynamic memory in many places. `std::array<DataT, Nr> coefficent` would be a more sensible choice.

Reply:
You are right. The grid size could also be just a member in the different classes and then one could use a std::vector everywhere, which would make the code a lot easier. But at the beginning I thought it would make sense to use arrays instead of vectors and to ensure that one uses the correct grid sizes (there are only specific grid sizes which can be used) and that these classes can only be used with the same grid size. I might remove the templates in the future, but this would require some work...

Indeed it makes sense to use an array there`std::array<DataT, Nr> coefficent`, but since this vector is passed to several member functions and in a different function in the code the same coefficient array is not depending on the templated grid size `std::vector<DataT> coefficient1(tnRRow)` it has to be a vector there. I could probably do some additional templating, to use vectors and arrays where appropriate.